src/js/components/TruncatingLabel.jsx

```diff
@@ -0,0 +1,225 @@
+// @flow
+
+import React, {Component, PropTypes} from 'react';
+import ReactDOM from 'react-dom';
+
+//--------------------------------------
+//  Safety constants
+//--------------------------------------
+
+const MINLENGTH = 5; // Minimum size of cut-down text
+const MAXLOOPS = 50; // Max no of iterations attempting to find the correct size text
+
+//--------------------------------------
+//  Flow types
+//--------------------------------------
+
+// Render lifecycle
+const RS_MEASURE = 'measure'; // Mounted, text/props changed, measurement needed.
+const RS_FLUID = 'fluid'; // Text too big, in the process of trimming it down
+const RS_STABLE = 'stable'; // Done measuring, until props change
+
+export const RenderStateValues = {
+    [RS_MEASURE]: RS_MEASURE,
+    [RS_FLUID]: RS_FLUID,
+    [RS_STABLE]: RS_STABLE,
+};
+
+type RenderState = $Keys<typeof RenderStateValues>;
+
+// react.Component.props
+type Props = {
+    children?: string,
+    style?: Object,
+    className?: string
+};
+
+/**
+ * Multi-line label that will truncate with ellipses
+ *
+ * Use with a set width + height (or maxWidth / maxHeight) to get any use from it :D
+ */
+export class TruncatingLabel extends Component {
+
+    //--------------------------------------
+    //  Component state / lifecycle
+    //--------------------------------------
+
+    props: Props;
+
+    completeText: string; // Unabridged plain text content
+    innerText: string; // Current innerText of element - includes possible ellipses
+    renderState: RenderState; // Internal rendering lifecycle state
+    checkSizeRequest: number; // window.requestAnimationFrame handle
+
+    //--------------------------------------
+    //  Binary search state
+    //--------------------------------------
+
+    textCutoffLength: number; // Last count used to truncate completeText
+    longestGood: number; // Length of the longest truncated text that fits
+    shortestBad: number; // Length of the shortest truncated text that does not fit
+    loopCount: number; // to avoid infinite iteration
+
+    //--------------------------------------
+    //  React Lifecycle
+    //--------------------------------------
+
+    componentWillMount() {
+        this.handleProps(this.props);
+    }
+
+    componentWillReceiveProps(nextProps: Props) {
+        this.handleProps(nextProps);
+    }
+
+    componentDidMount() {
+        this.invalidateSize();
+    }
+
+    componentDidUpdate() {
+        this.invalidateSize();
+    }
+
+    componentWillUnmount() {
+        if (this.checkSizeRequest) {
+            cancelAnimationFrame(this.checkSizeRequest);
+            this.checkSizeRequest = 0;
+        }
+    }
+
+    //--------------------------------------
+    //  Render
+    //--------------------------------------
+
+    render() {
+        const {
+            style = {},
+            className = ''
+        } = this.props;
+
+        const mergedStyle = {
```

Comment:
All components should reflect incoming className and style props onto the divs they generate.